=== modified file 'vm/complete/retro.py'
--- vm/complete/retro.py	2012-06-20 10:58:45 +0000
+++ vm/complete/retro.py	2012-10-13 16:34:45 +0000
@@ -223,8 +223,21 @@
   ports = [0] * 12
   files = [0] * 8
 
+  # -- trace file patch --
+  # if you want to test your ngaro implementation, you can apply this
+  # patch to retro.py and it will log the interpreter state before 
+  # each tick of the vm. ( the idea is you do the same in your own
+  # implementation and diff the files. )
+  #
+  # The records are delimited by newlines and "" chr( 31 ),
+  # the ascii/unicode "unit separator" character.
+  trace = open( 'python.log', 'a' )
   while ip < ext:
     opcode = memory[ip]
+    trace.write( '%s%s%s%s\n' %
+                ( ip, opcode,
+                  ' '.join( str(i) for i in stack ),
+                  ' '.join( str(i) for i in address )))
 
     # If opcode is not one we know,
     # jump past it. We have 30-opcodes.
@@ -418,7 +431,8 @@
     t[0] = 0
     termios.tcsetattr(sys.stdin.fileno(), termios.TCSANOW, t)
   try: process( memory, inputs )
-  except: pass
+  except:
+      raise
   if set_termio:
     termios.tcsetattr(sys.stdin.fileno(), termios.TCSANOW, old)
 

